tests/communication/test_decorator.py

```diff
@@ -0,0 +1,94 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+
+from concurrent.futures import ThreadPoolExecutor, as_completed
+import os
+import redis
+import subprocess
+import sys
+import time
+import threading
+import unittest
+
+from maro.communication import Proxy, SessionMessage, dist
+from utils import get_random_port
+
+
+def proxy_generator(component_type, redis_port):
+    if component_type == "receiver":
+        proxy = Proxy(group_name="proxy_unit_test",
+                      component_type="receiver",
+                      expected_peers={"sender": 1},
+                      redis_address=("localhost", redis_port),
+                      log_enable=False)
+    elif component_type == "sender":
+        proxy = Proxy(group_name="proxy_unit_test",
+                      component_type="sender",
+                      expected_peers={"receiver": 1},
+                      redis_address=("localhost", redis_port),
+                      log_enable=False)
+    return proxy
+
+
+def handler_function(that, proxy, message):
+    replied_payload = {"counter": message.payload["counter"] + 1}
+    proxy.reply(message, payload=replied_payload)
+    sys.exit(0)
+
+
+def lunch_receiver(proxy, handler_dict):
+    @dist(proxy, handler_dict)
+    class Receiver:
+        def __init__(self):
+            pass
+
+    receiver = Receiver()
+    receiver.launch()
+
+
+@unittest.skipUnless(os.environ.get("test_with_redis", False), "require redis")
+class TestDecorator(unittest.TestCase):
+    @classmethod
+    def setUpClass(cls):
+        print(f"The dist decorator unit test start!")
+        # Prepare Redis.
+        random_port = get_random_port()
+        cls.redis_process = subprocess.Popen(["redis-server", "--port", str(random_port), "--daemonize yes"])
+        cls.redis_process.wait()
+
+        # Prepare proxy.
+        proxy_type_list = ["receiver", "sender"]
+        with ThreadPoolExecutor(max_workers=2) as executor:
+            all_task = [executor.submit(proxy_generator, proxy_type, random_port) for proxy_type in proxy_type_list]
+
+            for task in as_completed(all_task):
+                result = task.result()
+                if "receiver" in result.component_name:
+                    cls.receiver_proxy = result
```

Comment:
why must cls, self doesn't work? also for others

Reply:
self vs cls
The difference between the keywords self and cls reside only in the method type. If the created method is an instance method then the reserved word self has to be used, but if the method is a class method then the keyword cls must be used. Finally, if the method is a static method then none of those words will be used because as said before, static methods are self-contained and do not have access to the instance or class variables nor to the instance or class methods.

Reply:
https://levelup.gitconnected.com/method-types-in-python-2c95d46281cd

Reply:
![image](https://user-images.githubusercontent.com/6368652/95674412-a7c2ad80-0be2-11eb-844e-a0fc785f90b6.png)

but why you use self.sender_proxy not cls.sender_proxy????